### ISSUE TITLE
Lore incorrectly throws away text nodes which are all whitespace, omitting them from the output document

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8049]
|<img alt="Nafai's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| Nafai reported|
|-|-|
|Trac ID|trac#414|
|Type|defect|
|Created|2003-11-25 11:00:12Z|
|Branch|https://github.com/twisted/twisted/tree/whitespace-preservation-414-3|

If a lore input document contains a fragment such as
```
<b>hello</b> <i>world</i>
```
then the output document will contain a fragment such as
```
<b>hello</b><i>world</i>
```
This is incorrect, and the output should be the same as the input in this case: nodes which are entirely whitespace are significant.

The root cause of this problem is that microdom erroneously throws away whitespace nodes.  The current proposed fix is to [ticket:3560 stop using microdom for lore, and to replace it with minidom], which implements correct handling of whitespace in XML documents.

<details><summary>Searchable metadata</summary>

```
trac-id__414 414
type__defect defect
reporter__Nafai Nafai
priority__normal normal
milestone__ 
branch__branches_whitespace_preservation_414_3 branches/whitespace-preservation-414-3
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__lore lore
keywords__ 
time__1069758012000000 1069758012000000
changetime__1469511682290030 1469511682290030
version__ 
owner__ 
cc__glyph cc__Nafai cc__spiv cc__itamarst cc__moshez cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="Nafai's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Nafai commented|
|-|-|

```
#!html
<pre>
For example, look at the Components howto and look for this:
&lt;q>smart&lt;/q> &lt;code>HairDryer&lt;/code>

And then look at the corresponding output
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)|@moshez commented|
|-|-|

```
#!html
<pre>
This is a problem in MicroDOMParser: it ignores gotText if it's all space
and the element is not shouldPreserveSpace. Perhaps it would be better
in that case to pretend we got exactly one space:

 def gotText(....)
     if ...:
         ....
     else:
+        self.gotStandAlone(Text, ' ')
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
empty space is dropped in order to up performance.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
Performance without correctness is not good. This is a bug. XML documents are usually space-
insensitive, but that only means you can translate multiple spaces into one space, not eat them entirely.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun removed owner|
|-|-|

I agree with James.

The branch whitespace-preservation-414 removes a test for the invalid behavior, adds a couple tests for the valid behavior, and changes the parser to collapse multiple spaces into one space, rather than eating them entirely.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @exarkun|
|-|-|

There are some failing tests in lore still which I guess I'll need to fix.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

This is pretty hard.  It's not actually a high priority for me now.  I hope this ticket can be resolved someday, but it'll probably be a while before I get back to it again.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [25530]) Branching to 'whitespace-preservation-414-3'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set status to `closed`|
|-|-|

The code in the branch now doesn't fix microdom, it makes lore not use microdom anymore.  This resolves the functional issue which prompted the creation of this ticket, but it doesn't fix microdom.  Since microdom's purpose is to be an XML library which doesn't change, changing it is probably a bad idea, so let's not.  Instead, let's encourage people to not use it.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph commented|
|-|-|

For what it's worth, microdom's purpose was to be an XML DOM which didn't change _so that lore would stop breaking_ when people installed different versions of the XML DOM implementation.  I think it's OK for it to change a little bit to make lore _less_ broken.

On the other hand, it's OK if it stops changing because it just isn't worth the effort to maintain it; switching to microdom is an OK resolution to the lore issue here.  Is there a new ticket for that issue, though, the lore bug (which the branch now fixes) rather than the microdom bug?  It sounds like something should be in review, but I don't see any other tickets mentioned.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph commented|
|-|-|

Or should I reopen and change the component to 'lore'?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

I filed [#3560](https://github.com/twisted/twisted/issues/3560) and put the code into a branch for that ticket.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph commented|
|-|-|

I thought about this for a while, and I think that the bug being reported here is actually in lore, not microdom.  The component said "web" but the summary refers to .xhtml files, which are not processed by microdom directly, they are loaded by lore.  I am guessing this component was chosen because the microdom and lore issues were (understandably, but incorrectly) conflated.

This ticket itself isn't too interesting, but it's an interesting exmplar of a particular type of debugging process pattern: a functional issue is discovered, the underlying library is discovered to be at fault, the underlying library can't or shouldn't be fixed trivially (due to compatibility concerns) and the functional issue gets dismissed because the underlying library issue will be fixed a different way.  In this case, we've actually committed to fix the functional issue but the status of this ticket is misleading.

I think that in general we should try to keep tickets which describe functional issues open until the issue is actually fixed.  Underlying library problems, when they are discovered to be distinct from the functional issue that they are the root cause of, should be shuffled off into separate tickets.  To make it clear here I've refined the summary and added a description explaining the behavior in detail.

In this case I'm not shuffling the library issue off into another ticket because I don't think anybody wants to fix it :).  I briefly considered fixing it myself, but I should probably find better things to do.

The branch in [#3560](https://github.com/twisted/twisted/issues/3560) is one strategy for fixing the functional issue in lore, but we should leave this ticket open in case someone (say, a contributor to thinkcspy) is interested in tracking the resolution of this problem, in the sense that they want to know what Twisted version actually shipped with a fixed `lore`.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set status to `reopened`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set status to `closed`|
|-|-|

(In [26264]) Merge minidom-lore-3560-4

Author: exarkun
Reviewer: radix, glyph
Fixes: [#3560](https://github.com/twisted/twisted/issues/3560), [#8049](https://github.com/twisted/twisted/issues/8049)

Switch Lore from using `twisted.web.microdom` to using `xml.dom`.

This explicitly makes the default Lore input document DTD XHTML1-Strict,
but allows either that DTD or XHTML1-Transitional (but not XHTML1-Frameset
or any other DTD).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_17"></a>|GitHub commented|
|-|-|

In changeset 2d4b239e897181d085b3c155421042ad05f83529
```
#!CommitTicketReference repository="" revision="2d4b239e897181d085b3c155421042ad05f83529"
Merge pull request #414 from twisted/8675-einarfd-ssh-filetransfer-2

Author: einarfd
Reviewer: rodrigc
Fixes: #8675

Port twisted.conch.ssh.filetranfer to Python 3
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_18"></a>|@glyph commented|
|-|-|

This pull request namespace conflict is hilarious but maybe we should start thinking about separating them somehow.
